### PR TITLE
feat: Assignment 10 Error Handling Initial Commit

### PR DESCRIPTION
**Pull request by Subhasri-V-P**
**Background:**
To understand Exception Handling and handle expected and unexpected errors.

**Implementation:**
1. Implemented try/catch/finally block to handle `DivisionByZeroException`.
2. Implemented try/catch/finally block to handle `OutOfRangeException` and a custom message exception is thrown.
3. Custom `InvalidUserInputException` class is created and invalid input for array elements are handled by throwing `InvalidUserInputException` custom exception.
4. Globally handled unexpected exceptions using `UnhandledExceptionEventHandler` and **HandleUnexpectedExceptions()** function.
5. Concepts of stack trace are explored and the printed stack trace is understood.

### CODE REVIEW

File: src/ErrorHandling/Program.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        /// <summary>
+        /// AddIntegersToArray() function adds integers got from the user in the array
+        /// </summary>
+        public static void AddIntegersToArray()
+        {
+            for (int i = 0; i < SizeOfArray; i++)
+            {
+                Console.WriteLine("Enter input for array: ");
+                string? unparsedInput = Console.ReadLine();
+                try
```
Comment (Shaalini-Subrammanian) [inline src/ErrorHandling/Program.cs:118]: Why do we have so much try..catch blocks?, Try catch has been already implemented for this method at the top level i.e at the method call, so we can remove this inside the method.

Always the try catch should be handled at the top level for example before the method calls.

**Review by Shaalini-Subrammanian — CHANGES_REQUESTED** [review]

File: src/ErrorHandling/Program.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        /// <summary>
+        /// GetValidInput - To get valid integer input
+        /// </summary>
+        /// <param name="unparsedString">Unparsed string to parse into integer</param>
+        /// <returns>Returns valid input</returns>
+        public static int GetValidInput(string? unparsedString)
+        {
+            int intInput;
```
[the remark this answers is not in the document]
Comment (Subhasri-V-P) [reply src/ErrorHandling/Program.cs:76]: Changes as parsedInput

File: src/ErrorHandling/Program.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        /// <summary>
+        /// Main class - Entry point
+        /// </summary>
+        /// <param name="args">String array of arguments</param>
+        public static void Main(string[] args)
+        {
+            AppDomain.CurrentDomain.UnhandledException += new UnhandledExceptionEventHandler(HandleUnhandledException);
+            string? unparsedInput1;
+            Console.WriteLine("Enter numerator: ");
+            unparsedInput1 = Console.ReadLine();
```
[the remark this answers is not in the document]
Comment (Subhasri-V-P) [reply src/ErrorHandling/Program.cs:34]: refactored code for clean execution

File: src/ErrorHandling/Program.cs (hunk 1 above)
Comment (Subhasri-V-P) [reply src/ErrorHandling/Program.cs:118]: The assignment says we have to handle the exception and throw a new exception with custom message. So I guess this is what is expected to be done.

Comment (Shaalini-Subrammanian) [reply src/ErrorHandling/Program.cs:109]: But it's not a best practice to have try catch everywhere, We should have the try catch at the top level

Comment (Subhasri-V-P) [reply src/ErrorHandling/Program.cs]: The exceptions were handled at top level

**Review by Shaalini-Subrammanian — APPROVED** [review]

**Pull request merged**